inginious/frontend/pages/preferences/profile.py

```diff
@@ -54,20 +55,23 @@ def save_profile(self, userdata, data):
             msg = _("Passwords don't match !")
             return result, msg, error
         elif self.app.allow_registration and len(data["passwd"]) >= 6:
-            oldpasswd_hash = UserManager.hash_password(data["oldpasswd"])
-            passwd_hash = UserManager.hash_password(data["passwd"])
-
+            ph = PasswordHasher()
             match = {"username": self.user_manager.session_username()}
+
             if "password" in userdata:
-                match["password"] = oldpasswd_hash
+                user = self.database.users.find_one(match)
+                oldpasswd_hash_sha512 = UserManager.hash_password_sha512(data["oldpasswd"])
+                try:
+                    if oldpasswd_hash_sha512 == user["password"] or ph.verify(user["password"][9:], data["oldpasswd"]):
```

Comment:
Here you support retrocompatibility by just testing both algorithms. This may be a bit dangerous in case one incorrect password with the incorrect algorithm would match the stored hash (in this particular case that won't happen, but this remains something possible with same-length same-alphabet hashes)

I strongly recommend that we extract the appropriate method from what is prepended. In this case, you will just need to fetch the `methods[prefix]` hashing function and perform the check. If the verifying process is different for argon2id, then the dictionary can point to classes that implement both hashing and verifying methods.

Reply:
Modified UserManager.auth_user() to do so. 

Modified auth_user() method to allow it's use during password change, reducing duplicate code.